kbs/src/api_server.rs

```diff
@@ -0,0 +1,242 @@
+                        .map_err(|e| Error::HTTPSFailed { source: e })?,
+                )
+                .map_err(|e| Error::HTTPSFailed { source: e.into() })?;
+
+            return tls_server
+                .run()
+                .await
+                .map_err(|e| Error::ServerLaunchFailed { source: e.into() });
+        }
+
+        http_server
+            .bind(&http_config.sockets[..])
+            .map_err(|e| Error::HTTPFailed { source: e.into() })?
+            .run()
+            .await
+            .map_err(|e| Error::ServerLaunchFailed { source: e.into() })
+    }
+}
+
+/// APIs
+pub(crate) async fn api(
+    request: HttpRequest,
+    body: web::Bytes,
+    core: web::Data<ApiServer>,
+) -> Result<HttpResponse> {
+    let query = request.query_string();
+    let plugin_name = request
+        .match_info()
+        .get("plugin")
+        .ok_or(Error::IllegalAccessedPath {
+            path: request.path().to_string(),
+        })?;
+    let sub_path = request
+        .match_info()
+        .get("sub_path")
+        .ok_or(Error::IllegalAccessedPath {
+            path: request.path().to_string(),
+        })?;
+
+    let end_point = format!("{plugin_name}{sub_path}");
+
+    match plugin_name {
+        #[cfg(feature = "as")]
+        "auth" if request.method() == Method::POST => core
+            .attestation_service
+            .auth(&body)
+            .await
+            .map_err(From::from),
+        #[cfg(feature = "as")]
+        "attest" if request.method() == Method::POST => core
+            .attestation_service
+            .attest(&body, request)
+            .await
+            .map_err(From::from),
+        #[cfg(feature = "as")]
+        "attestation-policy" if request.method() == Method::POST => {
+            core.attestation_service.set_policy(&body).await?;
+
+            Ok(HttpResponse::Ok().finish())
+        }
+        "resource-policy" if request.method() == Method::POST => {
+            core.admin_auth.validate_auth(&request)?;
+            core.policy_engine.set_policy(&body).await?;
+
+            Ok(HttpResponse::Ok().finish())
+        }
+        "resource-policy" if request.method() == Method::GET => {
+            core.admin_auth.validate_auth(&request)?;
+            let policy = core.policy_engine.get_policy().await?;
+
+            Ok(HttpResponse::Ok().content_type("text/xml").body(policy))
+        }
+        plugin_name => {
+            let plugin = core
+                .plugin_manager
+                .get(plugin_name)
+                .ok_or(Error::PluginNotFound {
+                    plugin_name: plugin_name.to_string(),
+                })?;
+
+            let body = body.to_vec();
+            if plugin
+                .validate_auth(&body, query, sub_path, request.method())
+                .await?
+            {
+                // Plugin calls needs to be authorized by the admin auth
+                core.admin_auth.validate_auth(&request)?;
```

Comment:
I am wondering what would be the scope of the admin auth (or any other auth) for plugins. The way it is implemented now, the plugin could use the admin auth for anything.

I can think of some ways to have a better separation of duties in order to reduce damage if an auth gets compromised. E.g. we could have one auth only for plugins; or picking a crazy idea from the jar, we could have a wallet maintained by admin and its keys/auth could be selected for plugins via rego policies.

Anyways, the design proposed in the PR looks good to me. We can make adjustments later as needed.

Reply:
I am not quite sure about the design for now. Before this PR, we have `set resource`, `set policy` like APIs who should be accessed via so-called "user public key auth". Also, we have APIs like `get resource` for clients with Attestation tokens. This PR, I just bring them together and rename to "admin auth" who needs old "user public key auth" and those who need attestation token as auth -- which I think better to decribe the two different roles up to now -- admin and client.

Also, the new design makes it easier for developers to integrate any new APIs that needs to be filtered by "admin auth" without changing too many codes.

Your opintion considers this issue more comprehensively and inspiringly. I think in future we will meet some new demands that require more roles for KBS and we can do some more extension upon the design then.

Reply:
> I am not quite sure about the design for now.

Do you want to expand on that? We could try to improve the parts you are not sure or find a way to move on only with the parts we are good. Honestly, like other reviewers, I also see many good changes in the PR. The auth thing I mentioned before is not introduced here, it is just something that may hit us in the future as we evolve trustee.

I am finishing to implement the nebula plugin on top of this PR, next week I should be able to post it. If the PR is not merged until then :smile:, the nebula plugin could be used to help the discussion of lingering questions.

Reply:
So if I understand correctly, the idea is that it is the admin's policy that defines which API needs admin auth, not the code of plugin, right?

Reply:
It seems a good idea to make use of policies to control which APIs (or plugin services) need admin auth. The policy could be defined by the plugin or the admin, there are tradeoffs in both cases, but I like the current design.

We don't know yet, but in the future we may have a plugin that wants to provide multiple services. IIUC, the current design allows the plugin to determine which of the services require admin_auth. Additionally, it is up to the plugin to make that decision based a policy e.g. loaded from a config file. Each plugin has its own requirements.

`            if plugin
                .validate_auth(&body, query, sub_path, request.method())
                .await?
            {
`

The admin does not need to know much about the business of the plugin (which plugin service should require admin_auth). That seems good to me. Later, if we think that's not enough, we can extend the admin to override some of the plugin's decision (not with the same granularity).

Reply:
Ok. So my understanding is to keep the logic as-is. Thanks for the explaination and if you have any new suggestion please raise them up

Reply:
Yep, looks good to me. Thanks